src/main/java/dev/galacticraft/mod/api/block/PipeShapedBlock.java

```diff
@@ -0,0 +1,180 @@
+import net.minecraft.world.level.block.state.StateDefinition;
+import net.minecraft.world.level.block.state.StateHolder;
+import net.minecraft.world.phys.shapes.CollisionContext;
+import net.minecraft.world.phys.shapes.Shapes;
+import net.minecraft.world.phys.shapes.VoxelShape;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public abstract class PipeShapedBlock<BE extends BlockEntity & Connected> extends Block implements EntityBlock {
+    public final VoxelShape[] shapes;
+
+    protected PipeShapedBlock(float radius, BlockBehaviour.Properties properties) {
+        super(properties);
+        this.shapes = makeShapes(radius);
+    }
+
+    @Override
+    abstract public @Nullable BE newBlockEntity(BlockPos pos, BlockState state);
+
+    abstract public boolean canConnectTo(Level level, BlockPos thisPos, Direction direction, BlockPos neighborPos, BlockState thisState);
+
+    abstract protected void onConnectionChanged(Level level, BlockPos thisPos, Direction direction, BlockPos neighborPos);
+
+    protected boolean updateConnection(BlockState currentState, BlockPos pos, Direction side, BlockPos neighborPos, Level level) {
+        if (level.getBlockEntity(pos) instanceof Connected pipe) {
+            boolean canConnect = this.canConnectTo(level, pos, side, neighborPos, currentState);
+
+            BlockState neighborState = level.getBlockState(neighborPos);
+            if (neighborState.getBlock() instanceof PipeShapedBlock<?> neighbor) {
+                canConnect &= neighbor.canConnectTo(level, neighborPos, side.getOpposite(), pos, neighborState);
+            }
+
+            boolean currentlyConnected = pipe.getConnections()[side.get3DDataValue()];
+            pipe.getConnections()[side.get3DDataValue()] = canConnect;
+            level.blockEntityChanged(pos);
+            return canConnect != currentlyConnected;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public void setPlacedBy(Level level, BlockPos pos, BlockState state, @Nullable LivingEntity placer, ItemStack itemStack) {
+        super.setPlacedBy(level, pos, state, placer, itemStack);
+
+        for (Direction direction : Direction.values()) {
+            this.updateConnection(state, pos, direction, pos.relative(direction), level);
+        }
+    }
+
+    @Override
+    public void neighborChanged(BlockState state, Level level, BlockPos pos, Block block, BlockPos neighborPos, boolean notify) {
+        super.neighborChanged(state, level, pos, block, neighborPos, notify);
+
+        Direction direction = Direction.fromDelta(neighborPos.getX() - pos.getX(), neighborPos.getY() - pos.getY(), neighborPos.getZ() - pos.getZ());
+        if (direction == null)
+            return;
+
+        if (this.updateConnection(state, pos, direction, neighborPos, level)) {
+            level.updateNeighborsAtExceptFromFacing(pos, state.getBlock(), direction);
```

Comment:
I might be reading the code incorrectly, but this feels really suspect to me. Wouldn't we want to specifically update the direction that got changed and not all the other faces? (the opposite of this method call)

Reply:
Probably some leftover from me trying to get blocks to update correctly. I've removed both calls and everything still works.